src/RestrictedTokenFaucet.t.sol

```diff
@@ -0,0 +1,160 @@
+        token.mint(address(faucet), 1000000 ether);
+        user1 = address(new FaucetUser(token, faucet));
+        user2 = address(new FaucetUser(token, faucet));
+        self = address(this);
+    }
+
+    function testSetupPrecondition() public {
+        assertEq(faucet.wards(self), 1);
+        assertEq(faucet.owner(), self);
+        assertEq(faucet.wards(user1), 0);
+        assertEq(faucet.wards(user2), 0);
+    }
+
+    function testFail_gulp_no_auth() public {
+        FaucetUser(user2).doGulp();
+    }
+
+    function test_gulp_auth() public {
+        faucet.rely(user1);
+        assertEq(faucet.wards(user1), 1);
+        assertEq(FaucetUser(user1).doBalanceOf(), 0);
+        FaucetUser(user1).doGulp();
+        assertEq(FaucetUser(user1).doBalanceOf(), 20 ether);
+    }
+
+    function testFail_rely_notOwner() public {
+        FaucetUser(user1).doRely(address(123));
+    }
+
+    function testFail_deny_notOwner() public {
+        FaucetUser(user1).doDeny(address(this));
+    }
+
+    function test_gulp_multiple() public {
+        address payable[] memory addrs = new address payable[](4);
+        addrs[0] = address(123);
+        faucet.rely(addrs[0]);
+        addrs[1] = address(234);
+        faucet.rely(addrs[1]);
+        addrs[2] = address(567);
+        faucet.rely(addrs[2]);
+        addrs[3] = address(890);
+        faucet.rely(addrs[3]);
+        assertEq(token.balanceOf(address(123)), 0);
+        assertEq(token.balanceOf(address(234)), 0);
+        assertEq(token.balanceOf(address(567)), 0);
+        assertEq(token.balanceOf(address(890)), 0);
+        faucet.gulp(address(token), addrs);
+        assertEq(token.balanceOf(address(123)), 20 ether);
+        assertEq(token.balanceOf(address(234)), 20 ether);
+        assertEq(token.balanceOf(address(567)), 20 ether);
+        assertEq(token.balanceOf(address(890)), 20 ether);
+    }
+
+    function testFail_gulp_multiple() public {
+        address payable[] memory addrs = new address payable[](4);
+        addrs[0] = address(this); // already rely'ed
+        addrs[1] = address(234); // not rely'ed
+        faucet.gulp(address(token), addrs);
+    }
+
+    function testFail_gulpTwice() public {
+        faucet.gulp(address(token));
+        faucet.gulp(address(token));
+    }
+
+    function test_unDone() public {
+        assertEq(token.balanceOf(address(this)), 0);
+        faucet.gulp(address(token));
+        assertEq(token.balanceOf(address(this)), 20 ether);
+        assertTrue(faucet.done(address(this), address(token)));
+        faucet.unDone(address(this), address(token));
+        assertTrue(!faucet.done(address(this), address(token)));
+        faucet.gulp(address(token));
+        assertEq(token.balanceOf(address(this)), 40 ether);
+    }
+
+    function testFail_unDone_notOwner() public {
+        faucet.gulp(address(token));
+        assertTrue(faucet.done(address(this), address(token)));
+        FaucetUser(address(user1)).doUnDone(address(this));
```

Comment:
```suggestion
        FaucetUser(user1).doUnDone(address(this));
```

But you might instead want to save them as `FaucetUser`, as they seem to be used in that form way more often.